src/IdealizedSimulations/neverworld_simulation.jl

```diff
@@ -0,0 +1,298 @@
+# Neverworld
+#
+# Ingredients:
+#
+#   * Zonally-periodic domain with continental shelves on all boundaries except Southern Ocean
+#       * longitude = (0, 60)
+#       * 2 configurations in latitude
+#           - Half basin: latitude = (-70, 0)
+#           - Full basin: latitude = (-70, 70)
+#       * z to -4000m
+#       * Southern Ocean channel from -60 to -40 (with a ridge to -2000 m)
+#
+#   * Zonally-homogeneous wind stress with mid-latitude jet and trade winds
+#
+#   * Buoyancy 
+#       * restoring hot at the equator and cold at the poles (parabola, cosine, smooth step function)
+#       * equator-pole buoyancy differential: 0.06 (α * g * 30 ≈ 0.06 with α=2e-4, g=9.81)
+#       * exponential initial vertical stratification with N² = 6e-5 and decay scale h = 1000 m
+#           - eg bᵢ = N² * h * exp(z / h)
+#
+#   * Quadratic bottom drag with drag_coefficient = 1e-3
+
+using CubicSplines
+
+using Oceananigans
+using Oceananigans.Units
+using Oceananigans.Grids: xnode, ynode
+using Oceananigans.Operators: Δzᶜᶜᶜ
+using Oceananigans.Coriolis: ActiveCellEnstrophyConservingScheme
+using Oceananigans.TurbulenceClosures: CATKEVerticalDiffusivity
+
+using ClimaOcean: CubicSplineFunction,
+                  u_bottom_drag, u_immersed_bottom_drag,
+                  v_bottom_drag, v_immersed_bottom_drag
+                  
+using ..VerticalGrids: stretched_vertical_cell_interfaces
+
+const c = Center()
+const f = Face()
+
+#####
+##### Geometry
+#####
+
+struct Point{T}
+    x :: T
+    y :: T
+end
+
+struct LineSegment{T}
+    p₁ :: Point{T}
+    p₂ :: Point{T}
+end
+
+struct Line{T}
+    p₁ :: Point{T}
+    p₂ :: Point{T}
+end
+
+distance(p₁::Point, p₂::Point) = sqrt((p₁.x - p₂.x)^2 + (p₁.y - p₂.y)^2)
+
+# https://stackoverflow.com/questions/849211/shortest-distance-between-a-point-and-a-line-segment
+function distance(point::Point, line::LineSegment)
+    x, y = point.x, point.y
+    x₁, y₁ = line.p₁.x, line.p₁.y
+    x₂, y₂ = line.p₂.x, line.p₂.y
+
+    # Line segment vector components
+    ℓˣ = x₂ - x₁
+    ℓʸ = y₂ - y₁
+
+    # Fractional increment to closest segment point
```

Comment:
This comment does not mean anything for me. After spending a few minutes I see what's happening here. It's clever. The clamp is crucial.

Here's a more geometric-inspired code

```Julia
𝐱  = [x,  y]
𝐱₁ = [x₁, y₁]
𝐱₂ = [x₂, y₂]

𝐥 = 𝐱₂ - 𝐱₁ # vector of the line segment

# we compute the ratio of the projection of 𝐱 on the direction 𝐥 over the length of 
ϵ = dot(𝐱, 𝐥) / norm(𝐥)^2

# if 0<ϵ<1 -> the closest point to 𝐱 is within the line segment
# if ϵ<0 -> the closest point to 𝐱 is 𝐱₁
# if ϵ>1 -> the closest point to 𝐱 is 𝐱₂

ϵ = clamp(ϵ, 0, 1)

𝐱′ = 𝐱₁ + ϵ * 𝐥 # this is = 𝐱₁ for ϵ=0 and =𝐱₂ for ϵ=1

return norm(𝐱, 𝐱′)
```

Reply:
or perhaps this is cleaner

```Julia
using LinearAlgebra

𝐱  = [x,  y,  0]
𝐱₁ = [x₁, y₁, 0]
𝐱₂ = [x₂, y₂, 0]

𝐥 = 𝐱₂ - 𝐱₁ # vector of the line segment

distance = dot(𝐱, 𝐥) < 0 ? norm(𝐱₁ - 𝐱) :
           dot(𝐱, 𝐥) > dot(𝐥, 𝐥) ? norm(𝐱₂ - 𝐱) :
           abs(cross(𝐱 - 𝐱₁, 𝐥)) / norm(𝐥)

return distance
```

Reply:
I didn't test it, but this is what it should be based on geometry, right?

Reply:
> Basically ϵ is a quantity that determines whether one of the edge points are the closest ones to 𝐱.

That's what I was trying to convey... the closest point on the segment to the target point is parameterized by `ϵ`. It's the relative distance along the segment from x1 to x2 to the closest segment point. So if `ϵ` the relative distance is 0 (ie, the closest point is x1). If `ϵ` is 1 then the relative distance is the full length of the segment is x2. Does that make sense? That's what the comment says. I don't find the linear algebra easier to understand ...

Reply:
"Fractional increment to closest segment point": the location of the closest segment point, parameterized as the relative distance along the segment pointing from x1 to x2.

dot norm and cross are opaque to me :-(

Reply:
I was trying to avoid if statements because it confuses me... I thought this process was clear:

1. Find the point on the segment that's closest to the target point
2. Compute the distance from the closest segment point to the target point

Reply:
`distance(point, line)` uses the formula from wikipedia:

https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line#:~:text=In%20Euclidean%20geometry%2C%20the%20distance,nearest%20point%20on%20the%20line.

I do think your code for `distance(point, line_segment)` is reasonable. But I'd vote for retaining the original code for point to line (I don't think invoking the cross product formula is better).

Reply:
I think the original function is more performant though. It doesn't matter, but if we did need to compute on the fly then we would need that original function. So I guess if we want to change to a new version we should at least keep the original as a comment or something.

Reply:
Cool. Keep the code. No worries. :)

But just to be fair you say “you don’t like if statement but you use clamp which is a very sneaky if statement.

Reply:
> `distance(point, line)` uses the formula from wikipedia:
> 
> https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line#:~:text=In%20Euclidean%20geometry%2C%20the%20distance,nearest%20point%20on%20the%20line.
> 
> I do think your code for `distance(point, line_segment)` is reasonable. But I'd vote for retaining the original code for point to line (I don't think invoking the cross product formula is better).

This wikipedia article needs work. I’ll fix it. :)